File: java/sourceafis/src/sourceafis/matching/minutia/MinutiaPairing.java
package sourceafis.matching.minutia;

import sourceafis.general.DetailLogger;
import sourceafis.templates.Template;

public class MinutiaPairing implements Cloneable 
{
    PairInfo[] candidateIndex;
    PairInfo[] probeIndex;
    PairInfo[] pairList;
    int pairCount;

    public DetailLogger.Hook logger = DetailLogger.off;

    public int getCount() { 
    	return pairCount; 
    }
    public PairInfo getLastAdded() {   
    	return pairList[pairCount - 1]; 
    }
    public void selectProbe(Template probe)
    {
        probeIndex = new PairInfo[probe.minutiae.length];
        pairList = new PairInfo[probe.minutiae.length];
        for (int i = 0; i < pairList.length; ++i)
           pairList[i] = new PairInfo();
        pairCount = 0;
    }

    public void selectCandidate(Template candidate)
    {
        if (candidateIndex == null || candidateIndex.length < candidate.minutiae.length)
        	candidateIndex = new PairInfo[candidate.minutiae.length];
        else{
        	for (int i = 0; i < candidateIndex.length; ++i)
            candidateIndex[i] = null;
        }
    }

    public void Reset(MinutiaPair root)
    {
        for (int i = 0; i < pairCount; ++i)
        {
            candidateIndex[pairList[i].pair.candidate] = null;
            probeIndex[pairList[i].pair.probe] = null;
            pairList[i].supportingEdges = 0;
        }
        
        candidateIndex[root.candidate] = probeIndex[root.probe] = pairList[0];
        pairList[0].pair = root;
        pairCount = 1;
    }

    public void add(EdgePair edge)
    {
    	 candidateIndex[edge.neighbor.candidate] = probeIndex[edge.neighbor.probe] = pairList[pairCount];
         pairList[pairCount].pair = edge.neighbor;
         pairList[pairCount].reference = edge.reference;
         ++pairCount;
      }

    public PairInfo getByCandidate(int candidate)
    {
        return candidateIndex[candidate];
    }

    public PairInfo getByProbe(int probe)
    {
        return probeIndex[probe];
    }

    public boolean isProbePaired(int probe)
    {
        return probeIndex[probe] != null;
    }

    public boolean isCandidatePaired(int candidate)
    {
        return candidateIndex[candidate] != null;
    }

    public PairInfo getPair(int index)
    {
        return pairList[index];
    }

    public void addSupportByProbe(int probe)
    {
        ++probeIndex[probe].supportingEdges;
    }
    public void log() { logger.log(this); }
   
    public MinutiaPairing clone() {
    	MinutiaPairing clone = new MinutiaPairing();
    	clone.probeIndex = (PairInfo[])probeIndex.clone();
        clone.candidateIndex = (PairInfo[])candidateIndex.clone();
        clone.pairList = new PairInfo[pairList.length];
        for (int i = 0; i < pairList.length; ++i)
        	if (pairList[i] != null)
        		clone.pairList[i] = (PairInfo)pairList[i].clone();
        clone.pairCount = pairCount;
        return clone;
    }
}

File: SourceAFIS/Matching/Minutia/MinutiaPairing.cs
using System;
using System.Collections.Generic;
using System.Text;
using SourceAFIS.General;
using SourceAFIS.Templates;

namespace SourceAFIS.Matching.Minutia
{
    public sealed class MinutiaPairing : ICloneable
    {
        PairInfo[] CandidateIndex;
        PairInfo[] ProbeIndex;
        PairInfo[] PairList;
        int PairCount;

        public DetailLogger.Hook Logger = DetailLogger.Null;

        public int Count { get { return PairCount; } }
        public PairInfo LastAdded { get { return PairList[PairCount - 1]; } }

        public void SelectProbe(FingerprintTemplate probe)
        {
            ProbeIndex = new PairInfo[probe.Minutiae.Count];
            PairList = new PairInfo[probe.Minutiae.Count];
            for (int i = 0; i < PairList.Length; ++i)
                PairList[i] = new PairInfo();
            PairCount = 0;
        }

        public void SelectCandidate(FingerprintTemplate candidate)
        {
            if (CandidateIndex == null || CandidateIndex.Length < candidate.Minutiae.Count)
                CandidateIndex = new PairInfo[candidate.Minutiae.Count];
            else
            {
                for (int i = 0; i < CandidateIndex.Length; ++i)
                    CandidateIndex[i] = null;
            }
        }

        public void Reset(MinutiaPair root)
        {
            for (int i = 0; i < PairCount; ++i)
            {
                CandidateIndex[PairList[i].Pair.Candidate] = null;
                ProbeIndex[PairList[i].Pair.Probe] = null;
                PairList[i].SupportingEdges = 0;
            }
            CandidateIndex[root.Candidate] = ProbeIndex[root.Probe] = PairList[0];
            PairList[0].Pair = root;
            PairCount = 1;
        }

        public void Add(EdgePair edge)
        {
            CandidateIndex[edge.Neighbor.Candidate] = ProbeIndex[edge.Neighbor.Probe] = PairList[PairCount];
            PairList[PairCount].Pair = edge.Neighbor;
            PairList[PairCount].Reference = edge.Reference;
            ++PairCount;
        }

        public PairInfo GetByCandidate(int candidate)
        {
            return CandidateIndex[candidate];
        }

        public PairInfo GetByProbe(int probe)
        {
            return ProbeIndex[probe];
        }

        public bool IsProbePaired(int probe)
        {
            return ProbeIndex[probe] != null;
        }

        public bool IsCandidatePaired(int candidate)
        {
            return CandidateIndex[candidate] != null;
        }

        public PairInfo GetPair(int index)
        {
            return PairList[index];
        }

        public void AddSupportByProbe(int probe)
        {
            ++ProbeIndex[probe].SupportingEdges;
        }

        public void Log() { Logger.Log(this); }

        public object Clone()
        {
            MinutiaPairing clone = new MinutiaPairing();
            clone.ProbeIndex = (PairInfo[])ProbeIndex.Clone();
            clone.CandidateIndex = (PairInfo[])CandidateIndex.Clone();
            clone.PairList = (PairInfo[])PairList.CloneItems();
            clone.PairCount = PairCount;
            return clone;
        }
    }
}
